### DISCUSSION TITLE
Problems using IntelliJ for Play 3.0 / Scala 3 / Pekko

### DISCUSSION THREAD

**Discussion by hertg** [playframework/playframework#12205] [Q&A]
Does anybody else have the issue where IntelliJ is unable to properly highlight syntax or has issues recognizing imported packages after upgrading to Play 3.0?

![image](https://github.com/playframework/playframework/assets/3533648/81922e83-9d1d-4f6c-b4a5-89bd831fa63f)

Curiously, when I `Ctrl` + click on the `org.apache.pekko.actor` import package, it jumps to the external library `org.apache.pekko:pekko-actor-typed`. No `AbstractActor` exists in this package.

![image](https://github.com/playframework/playframework/assets/3533648/63218762-0401-4bcf-b9d3-33ba1858c3ea)

Shouldn't it jump to this dependency?

![image](https://github.com/playframework/playframework/assets/3533648/613b59f6-40a5-4ed1-8530-1d3b4d4636ce)

---

I'm using: 
- Latest IntelliJ Ultimate Edition
- Latest Scala Plugin (using nightly channel didn't help)
- sbt 1.9.7
- scala 3.3.1
- play 3.0.0

**Comment by ihostage** [comment]
Did you try `File -> Repair IDE`? 
Do examples from https://github.com/playframework/play-samples/ have same problem to you?

**Reply by hertg** [reply]
I tried out the `play-java-pekko-cluster-example`, and it didn't have this issue. So I compared the build configurations, and it seems to be an issue related to the scala version specified.

The following works:

```sbt
crossScalaVersions := Seq("2.13.12", "3.3.1")
scalaVersion := crossScalaVersions.value.head
```

This does not:

```sbt
scalaVersion := "3.3.1"
```

---

Please note, the sbt compilation and running the project always worked. The problem I'm having is simply that IntelliJ shows errors. Is this an issue with IntelliJ not properly supporting Scala 3 projects?

**Comment by hertg** [comment]
Here are some more troubles I had to address, just so they are documented somewhere, maybe it helps someone (or myself) in the future.

After opening my **play 3.0 / scala 3 / sbt 1.9.7 / jdk 21** project in IntelliJ, I have to manually do the following:
- Open `sbt settings`, and change JRE to `21` (it defaulted to `8` for some reason)
- Open Project Settings and set the SDK version to `21` (here it actually took my system default)
- When adding a run configuration for Play Framework, manually remove the `-XX:+CMSClassUnloadingEnabled` argument that is added by default but doesn't seem to exist in Java 21.
- To make the [sbt-dotenv](https://github.com/Philippus/sbt-dotenv) plugin work, I had to add JVM arguments after upgrading from JDK 8 to JDK 21 (also see [here](https://github.com/Philippus/sbt-dotenv?tab=readme-ov-file#illegal-reflective-access-warnings-and-exceptions))
  ```
  --add-opens=java.base/java.util=ALL-UNNAMED
  --add-opens=java.base/java.lang=ALL-UNNAMED
  ```

**Comment by mkurz** [comment]
@hertg Seems you are not alone: https://discord.com/channels/931647755942776882/931955815479402546/1174312904397434951
Seems like a IntelliJ problem (or likely it's Scala plugin)

**Reply by mkurz** [reply]
For the record - IntelliJ ticket here: https://youtrack.jetbrains.com/issue/SCL-21800/Highlightning-of-Pekko-classes-works-for-Scala-2.13.12-but-not-for-Scala-3.3.1

**Comment by hertg** [comment]
**Update**

I've upgraded my project to Scala 3 today and found that this issue has been resolved. However, there is another issue where in my play-java project most Java source files had syntax highlighting disabled. I found the following bug report at IntelliJ: 

https://youtrack.jetbrains.com/issue/SCL-19340/Highlighting-disabled-for-Java-files-in-Java-Scala-3-mixed-project

In order to fix this issue, one can disable the `scala.compiler.highlighting.suppress.java.highlighting` setting in the IntelliJ registry, see: https://youtrack.jetbrains.com/issue/SCL-19340/Highlighting-disabled-for-Java-files-in-Java-Scala-3-mixed-project#focus=Comments-27-6721077.0-0

**Comment by hertg** [comment]
@mkurz Hi Matthias, quick question, are there issues to be expected in moving to a Scala 3 project (no cross-version), or should this work in your opinion? I see that in the sample projects, you are still using cross-versioning with Scala 2.13. Is there any reason for that?

https://github.com/playframework/play-samples/blob/31d7ba7e0f4704372e51c38c79b32c7c0981234c/play-java-starter-example/build.sbt#L7-L8

I tried to go full Scala 3, and it seemed to work quite well, but sometimes I get random `NoClassDefFound` exceptions, even for classes that I have created myself. I don't know if it's related to the Scala 3 update though, maybe I've messed up something else.

**Reply by mkurz** [reply]
> are there issues to be expected in moving to a Scala 3 project (no cross-version), or should this work in your opinion?

No, there should not be any show stoppers, specially if you use Pekko. There might be issues with Scala 3 itself, but not with Play. Just read through https://www.playframework.com/documentation/3.0.x/Scala3Migration - there is only one thing, the guice regression at the bottom of the page that people ran into, but again, this is not Play's fault.

> I see that in the sample projects, you are still using cross-versioning with Scala 2.13. Is there any reason for that?

The reason is that we keep offering Play artifacts for Scala 2.13 and 3 in parallel, both are first class citizens. Scala 2.13 will be around for a looong time to come, so there is no reason to drop it anytime soon. There was a discussion recently about Scala 2.13 EOL [here](https://contributors.scala-lang.org/t/2-13-eol-and-lts-planning/6664), and and Seth from the Scala 2.13 team made clear that _"releases will continue indefinitely, as long into the future as anyone can possibly anticipate, certainly for years to come and almost certainly for a decade or more. There is no plan, and not even any desire, to EOL 2.13 at all."_
So that's why we will continue cross build. Also because there will be many many big projects out there that would take years to migrate to Scala 3.

> I tried to go full Scala 3, and it seemed to work quite well, but sometimes I get random NoClassDefFound exceptions, even for classes that I have created myself. I don't know if it's related to the Scala 3 update though, maybe I've messed up something else.

If you run into that next time please report that somewhere, so we or the Scala team can figure out what's going on.

**Reply by hertg** [reply]
Thanks for your quick response! I see, that sounds reasonable. But if I am starting a new project that focuses on using a minimum amount of external dependencies, there isn't really a benefit of doing cross-compilation, right? (Ignoring the fact that IntelliJ still has issues with Scala 3)

Unfortunately, I don't have any idea what to report here, because its one of those issues that is almost impossible to reproduce. I have recently enabled the `Zinc` instead of `IDEA` Scala Compiler in IntelliJ, in the hope that this would make IntelliJ less laggy. It didn't help. I just checked if that was the culprit of the sudden increase in NoClassDefFound fatal errors, so I switched it back to `IDEA`, ran a clean, rebuilt the project, and that didn't seem to help.

What's funny is that I just started the project in the sbt shell directly, and I haven't gotten any NoClassDef errors in the last 5 minutes. Usually I start my project with the IntelliJ "Play Framework" run configuration... Could it be that there's an issue there?

**Reply by mkurz** [reply]
> . But if I am starting a new project that focuses on using a minimum amount of external dependencies, there isn't really a benefit of doing cross-compilation, right?

No, for you there is absolutely no benefit to cross built your projects for Scala 2 and 3. We do that just to test if all the samples work in both Scala 2 and 3. "End users" (like you) just choose to work in Scala 2 _or_ Scala 3. Unless you write libraries used by others and you want provide support for Scala 2 and 3 there is absolutely no reason to cross build your projects. (In theory your company could have Scala 2 and Scala 3 projects and these projects could share a library, then you would have to cross build that library).

> Could it be that there's an issue there?

If you are talking about Scala 3 within IntelliJ that could be it. I think the Scala 3 plugin in IntelliJ is not 100% reliable yet. But they are working on it.

**Reply by hertg** [reply]
Oh, that clears up some things for me, thanks for pointing it out. Yeah, I've used cross-compilation before, for libraries that had to be cross-compiled between 2.12 and 2.13. The cross-compilation in the sample projects threw me off a bit (especially with 2.13 / 3, because I'm less familiar with their compatability) so I assumed that there is a need for that in non-library projects when using differing libraries.

I'm inclined to blame it on IntelliJ now, it seems that after I started it in the sbt shell, it now works with the Play App run configuration too... Weird. I'll make sure to note it somewhere if I figure out more about it.

Thanks for your responses, even though Play clearly isn't at fault here, appreciate your help.

**Reply by mkurz** [reply]
You are welcome :+1:

**Reply by hertg** [reply]
Is there a reason to wait on updating sbt to `1.10`? I'm currently on `1.9.9.`.

*sidenote: the sbt shell also caused the NoClassDefFound error, I haven't seen it yet when running manually from the terminal. I also just updated my Scala plugin from `2024.1.630` to `2024.1.653` and hope that it helps, will start it via IntelliJ again. If I continue to get errors, I might have to downgrade to 2.13.*

**Reply by mkurz** [reply]
> Is there a reason to wait on updating sbt to `1.10`? I'm currently on `1.9.9.`.

No, absolutely no reason to wait. You can use 1.10 right now.

**Reply by hertg** [reply]
**Update**: It seems like it's working now, but every time I start the project, I get tons of Scala compiler errors **on the first request**. All of them seem to be related to imports in my twirl templates. Upon initiating a second request, it starts compiling again, and then everything works fine :thinking: 

Play `3.0.4-M1` / sbt 1.10.0 / JDK 21.0.3 / Scala 3.3.3

---

*heavily truncated and redacted*
```txt
[info] compiling 143 Scala sources and 1 Java source to /mnt/data/repositories/myproject/target/scala-3.3.3/classes ...
[error] /mnt/data/repositories/myproject/app/org/example/myproject/account/views/webauthn_rename.scala.html:1:1: value utils is not a member of org.example.myproject.template
[error] @import org.example.myproject.account.models.FormFidoName
[error] ^
[error] /mnt/data/repositories/myproject/app/org/example/myproject/account/views/setup_webauthn_success.scala.html:1:1: value utils is not a member of org.example.myproject.template
[error] @import org.example.myproject.account.models.FormFidoName
[error] ^
[error] /mnt/data/repositories/myproject/app/org/example/myproject/account/views/setup_webauthn_start.scala.html:1:1: value utils is not a member of org.example.myproject.template
[error] @import org.example.myproject.account.views.html.account_layout
[error] ^
...

An unhandled exception was thrown in the compiler.
  Please file a crash report here:
  https://github.com/lampepfl/dotty/issues/new/choose

     while compiling: <no file>
        during phase: parser
                mode: Mode()
     library version: version 2.13.12
    compiler version: version 3.3.3
...

[error] ## Exception when compiling 679 sources to /mnt/data/repositories/myproject/target/scala-3.3.3/classes
[error] java.lang.StackOverflowError
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3191)
[error] dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3303)
[error] dotty.tools.dotc.typer.Namer.typedAheadExpr$$anonfun$1(Namer.scala:1656)
[error] dotty.tools.dotc.typer.Namer.typedAhead(Namer.scala:1646)
[error] dotty.tools.dotc.typer.Namer.typedAheadExpr(Namer.scala:1656)
[error] dotty.tools.dotc.typer.Namer$Completer.$anonfun$15(Namer.scala:794)
...
[error] (Compile / compileIncremental) java.lang.StackOverflowError
ERROR o.a.p.a.ActorSystemImpl - Internal server error, sending 500 response
scala.MatchError: java.lang.StackOverflowError (of class java.lang.StackOverflowError)
	at play.sbt.run.PlayReload$.$anonfun$taskFailureHandler$8(PlayReload.scala:92)
	at scala.Option.map(Option.scala:230)
	at play.sbt.run.PlayReload$.taskFailureHandler(PlayReload.scala:92)
	at play.sbt.run.PlayReload$.$anonfun$compile$3(PlayReload.scala:144)
	at scala.util.Either$LeftProjection.map(Either.scala:573)
	at play.sbt.run.PlayReload$.compile(PlayReload.scala:144)
	at sbt.PlayRun$.$anonfun$playRunTask$5(PlayRun.scala:99)
	at play.runsupport.Reloader.$anonfun$reload$1(Reloader.scala:495)
	at play.runsupport.Reloader$$anon$1.run(Reloader.scala:58)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:400)
```

**Comment by mkurz** [comment]
> value utils is not a member of org.example.myproject.template

> @import org.example.myproject.account.models.FormFidoName

Are you using annotations somewhere? Or some other annotation processor?

Is FormFidoName a Java class?

**Reply by hertg** [reply]
Good point, yes I do use annotation processing. `FormFidoName` is a Java class.

<details>
<summary>FormFidoName Source</summary>

```java
import play.data.validation.Constraints;
import play.data.validation.ValidationError;

import java.util.ArrayList;
import java.util.List;

@Constraints.ValidateWithPayload
public class FormFidoName implements Constraints.ValidatableWithPayload<List<ValidationError>> {

    @Constraints.Required(message = "INPUT_FIELD_REQUIRED")
    private String name;

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    @Override
    public List<ValidationError> validate(Constraints.ValidationPayload payload) {
        return new ArrayList<>();
    }

}
```
</details>

**Reply by mkurz** [reply]
hm.. this does look ok actually.
Maybe you can just look directly at the generated `webauthn_rename*.scala` file in your target folder to see which other imports ar generated in that file?
Also do you define `TwirlKeys.templateImports` somewhere?

**Reply by hertg** [reply]
I see that there's a mix of `_` and `*` that's used for wildcard imports, but I doubt that's the problem? I switched my imports to use the `*` everywhere.
Yes I do have that in my build.sbt, just for convenience, so that I don't have to manually import the things I use almost everywhere.

```scala
// contains some utility methods and the implicit val FieldConstructor
TwirlKeys.templateImports += "org.example.template.utils.TemplateHelper.*"
// Contains html components like Button, Icon, etc
TwirlKeys.templateImports += "org.example.template.models.*"
```

**Reply by mkurz** [reply]
> I see that there's a mix of `_` and `*` that's used for wildcard imports, but I doubt that's the problem? I switched my imports to use the `*` everywhere.

That should not matter, AFAIK current Scala 2.13.x and 3.3.3 releases can handle both out of the box.

> ```sbt
> // contains some utility methods and the implicit val FieldConstructor
> TwirlKeys.templateImports += "org.example.template.utils.TemplateHelper.*"
> ````

Did you actually mean 
```sbt
TwirlKeys.templateImports += "org.example.myproject.template.utils.TemplateHelper.*"
```
Because this would then fit the error message: `value utils is not a member of org.example.myproject.template`

**Reply by hertg** [reply]
Yes, that's what I meant.  Sorry, was redacting some stuff on the fly :smile:

**Reply by mkurz** [reply]
So, this then is the import that causes problems... Just to clarify, this happens only within Intellij or also on the cmdline when running vanilla sbt?

**Reply by hertg** [reply]
I just got a random `NoClassDef` exception again, for the first time this week. I think I'll downgrade to Scala 2.13 because the current IntelliJ issues are not worth it for me right now (It's also slooooow). Will probably try upgrading again at some later time, but since you mentioned that 2.13 won't be deprecated anytime soon, there's no hurry it seems :)

**Reply by mkurz** [reply]
Yes, since you use Play Java anyway (I guess), it does absolutely make no difference.

**Reply by hertg** [reply]
> So, this then is the import that causes problems... Just to clarify, this happens only within Intellij or also on the cmdline when running vanilla sbt?

Sorry, missed your comment. The issues I reported were all in IntelliJ, so I'm inclined to blame it on that. I haven't had them when running sbt in my terminal.

**Reply by hertg** [reply]
> Yes, since you use Play Java anyway (I guess), it does absolutely make no difference.

I use Scala in some places (like 5% of the code), but keep it to a minimum to keep it maintainable for Java devs. There were some things I was excited about in Scala 3, like the reworked implicit stuff and trait parameters for example. But I haven't gotten used to it yet, so I can live without those things for a while :smile:
